src/compiler/binder.ts

```diff
@@ -167,9 +176,45 @@ namespace ts {
                 case SyntaxKind.FunctionDeclaration:
                 case SyntaxKind.ClassDeclaration:
                     return node.flags & NodeFlags.Default ? "default" : undefined;
+                case SyntaxKind.JSDocFunctionType:
+                    return isJSDocConstructSignature(node) ? "__new" : "__call";
+                case SyntaxKind.Parameter:
+                    // Parameters with names are handled at the top of this function.  Parameters
+                    // without names can only come from JSDocFunctionTypes.
+                    Debug.assert(node.parent.kind === SyntaxKind.JSDocFunctionType);
+                    let functionType = <JSDocFunctionType>node.parent;
+                    let index = indexOf(functionType.parameters, node);
+                    return "p" + index;
+
+                case SyntaxKind.BinaryExpression:
+                    Debug.assert(isModuleExportsAssignment(node));
+                    return "__jsExports";
             }
         }
 
+        /**
+         * Returns the first invocation of an identifier 'require' with one argument,
+         * or undefined if that none exist.
+         */
+        function findCommonJSRequireCalls(node: Node): Node {

```

Comment:
This effectively does a full AST walk, which is expensive. We should just collect this information in the binder.